### PR TITLE
Express mongoose promises

### PR DESCRIPTION
**Pull request by hamishwillee**
This is the example code work to get the express tutorial to work with mongoose 7, as raised in https://github.com/mdn/content/issues/24972. 

Once this is complete we will have to update the content on MDN to match, and merge them at the same-ish time.
This includes replacing mentions of the async module in https://developer.mozilla.org/en-US/docs/Learn/Server-side/Express_Nodejs/mongoose#testing_%E2%80%94_create_some_items and https://developer.mozilla.org/en-US/docs/Learn/Server-side/Express_Nodejs/Displaying_data

The solution uses [express-async-handler](https://www.npmjs.com/package/express-async-handler) some simple middleware that allows you to make async/await calls in your express handlers, and then pushes any errors to the normal handler. An explanation of the logic to this can be found in this article: https://www.wisdomgeek.com/development/web-development/using-async-await-in-expressjs/

- [x] authorController
- [x] genreController
- [x] bookController
- [x] bookInstanceController
- [x] Fix up routes files etc to use const (or `let`  in rare cases
- [x] prettier everything

### CODE REVIEW

File: controllers/authorController.js [hunk 1]
```diff
@@ -1,269 +1,213 @@
...
-      );
-    }
-  },
-];
+var Author = require("../models/author");
+var async = require("async");
+var Book = require("../models/book");
+
+const { body, validationResult } = require("express-validator");
+const asyncHandler = require("express-async-handler");
```
Comment (hamishwillee) [inline controllers/authorController.js:6]: This requires the handler - method that catches your errors and sends them to normal express handler, and wraps rest of your code in await.

File: controllers/authorController.js [hunk 2]
```diff
@@ -1,269 +1,213 @@
...
+  res.render("author_list", {
+    title: "Author List",
+    author_list: authors,
+  });
+});
+
+// Display detail page for a specific Author.
+exports.author_detail = asyncHandler(async (req, res, next) => {
+  // Get details of author and all their books (in parallel)
+  const [author, authors_books] = await Promise.all([
```
Comment (hamishwillee) [inline controllers/authorController.js:20]: This is syntax for waiting for multiple promises and returning when all done. Need to explain this as we pulled out he async module.

**Comment by hamishwillee** [comment]
PR is ready to go, but keeping in draft until updates to tutorial are in place.

File: controllers/bookController.js [hunk 3]
```diff
@@ -1,114 +1,84 @@
...
+    numAuthors,
+    numGenres,
+  ] = await Promise.all([
+    Book.countDocuments({}),
+    BookInstance.countDocuments({}),
+    BookInstance.countDocuments({ status: "Available" }),
+    Author.countDocuments({}),
+    Author.countDocuments({}),
+    //Book.find({ author: req.params.id }, "title summary").exec(),
+  ]);
```
Comment (hamishwillee) [inline controllers/bookController.js:24]: Remove spurious line
```suggestion
    Author.countDocuments({}),
  ]);
```

File: controllers/bookController.js [hunk 4]
```diff
@@ -1,114 +1,84 @@
...
+    numAvailableBookInstances,
+    numAuthors,
+    numGenres,
+  ] = await Promise.all([
+    Book.countDocuments({}),
+    BookInstance.countDocuments({}),
+    BookInstance.countDocuments({ status: "Available" }),
+    Author.countDocuments({}),
+    Author.countDocuments({}),
+    //Book.find({ author: req.params.id }, "title summary").exec(),
```
Comment (hamishwillee) [inline controllers/bookController.js:23]:
```suggestion
```

**Comment by hamishwillee** [comment]
Ready to go now. Needs to be merged in parallel with https://github.com/mdn/content/pull/25899

**Review by teoli2003 — APPROVED** [review]
👍

**Comment by teoli2003** [comment]
@schalkneethling I can't merge here. Given that this code is to be kept in sync with what is in mdn/content, it would be nice if the same set of people could merge here as the ones that can merge at mdn/content. Thank you.

**Comment by hamishwillee** [comment]
Yes it would. I can merge and will do so, but please add @teoli2003 at least because he's the most likely reviewer of this content.

**Pull request merged**